New Prime Numbers Goodie

haojunsui opened the pull request:
**What does your Instant Answer do?**
It will give instant answer for prime numbers in given range.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Searches from online does not give primes in a given range, and it's super hard to get primes in a given range.

**What is the data source for your Instant Answer? (Provide a link if possible)**
I used Sieve of Eratosthenes algorithm for find prime numbers.

**Why did you choose this data source?**
It's easy and fast with finding small primes.

**Are there any other alternative (better) data sources?**
No.

**What are some example queries that trigger this Instant Answer?**
prime numbers between 13 and 325

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Cryptography.

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
RandomNumber

**Which existing Instant Answers will this one supercede/overlap with?**
No.

**Are you having any problems? Do you need our help with anything?**
No.

**Where did you hear about DuckDuckHack? (For first time contributors)**
School.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
<img width="1440" alt="screen shot 2015-10-17 at 2 18 52 pm" src="https://cloud.githubusercontent.com/assets/6809863/10560475/0acfada6-74da-11e5-96ee-7b5260ba1e37.png">


```
- [X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
- [X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
- [X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
- [X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
- [X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 10
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] IE 11
        - [] Microsoft edge
        
    - Windows 8
        - [X] Google Chrome
        - [X] Firefox
        - [] Opera
        - [] IE 10

    - Windows 7
        - [X] Google Chrome
        - [] Firefox
        - [] Opera
        - [] IE 8
        - [] IE 9
        - [] IE 10

    - Windows XP
        - [] IE 7
        - [] IE 8

    - Linux
        - [] Google Chrome
        - [X] Firefox
        
    - Mac OSX
        - [X] Google Chrome
        - [X] Firefox
        - [] Opera
        - [X] Safari

    - iOS (iPhone)
        - [] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - iOS (iPad)
        - [] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - Android
        - [] Firefox
        - [] Native Browser
        - [] Google Chrome
        - [] Opera

```

-----
IA Page: https://duck.co/ia/view/prime_numbers

daxtheduck commented:
@shjnyr Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please send an email to <quackslack@duckduckgo.com>.

>More Info: https://duck.co/duckduckhack/submission_and_review

Thanks!

mintsoft commented:
Hi @shjnyr we'll take a look at this shortly and give you some feedback. Something I've noticed immediately is that there's the new Goodie and a tutorial IA in this PR. Can you remove the IsAwesome?

mintsoft commented on lib/DDG/Goodie/PrimeNumber.pm:
```diff
@@ -0,0 +1,85 @@
+use strict;
+
+zci answer_type => "prime";
+zci is_cached   => 0;
```
This can be cached, as it doesn't have any random output

mintsoft commented on lib/DDG/Goodie/PrimeNumber.pm:
```diff
@@ -0,0 +1,85 @@
+            github => ["https://github.com/shjnyr", "shjnyr"],
+            twitter => ["https://twitter.com/Charles_Sui", "Haojun Sui"];
+
+# Triggers
```
You can remove these comments if you want, as they're largely for instruction

mintsoft commented on lib/DDG/Goodie/PrimeNumber.pm:
```diff
@@ -0,0 +1,85 @@
+# Triggers
+triggers start => "prime", "prime numbers";
+
+# Handle statement
```
You should only have `handle remainder` or `handle query_lc`

mintsoft commented on t/PrimeNumber.t:
```diff
@@ -0,0 +1,22 @@
+    'example query' => test_zci('query'),
+    # Try to include some examples of queries on which it might
+    # appear that your answer will trigger, but does not.
+    'bad example query' => undef,
```
Can you add some tests for expected output here? At the very minimum the `primary_example_queries` should have tests.

shjnyr commented:
I added test feature and updated the original code file. Thank you.

mintsoft commented:
@shjnyr Thanks for the updates, it could appear that there's now a least common multiple Goodie in this PR. Can you break this out into a separate PR?

mintsoft commented on t/PrimeNumber.t:
```diff
@@ -0,0 +1,22 @@
+            operation => 'Prime numbers between',
+            result    => qr/(\d{1,3}\,\s\s)+\d{1,3}/,
+        }
+    ),
```
Can you add some tests for pathological inputs? Negative numbers, queries like `prime numbers between hello and zebra`

moollaza commented on lib/DDG/Goodie/PrimeNumber.pm:
```diff
@@ -0,0 +1,91 @@
+    $end = 1000000000 if $end > 1000000000;
+    ($end, $start) = ($start, $end) if ($start > $end);
+    
+    my $s = ceil($start);
```
@shjnyr while i really appreciate you've written this code, I think it's best to use an existing CPAN module. Like we did for LCM, `Math::Prime::Util` has a method for detecting primes that's probably more performant. We can also load it with `try => GMP` and then we can take advantage of the C-Bindings in production.

moollaza replied:
We could leverage https://metacpan.org/pod/Math::Prime::Util#primes

Also, we could potentially expose https://metacpan.org/pod/Math::Prime::Util#is_prime here, or in another Goodie

moollaza commented on lib/DDG/Goodie/PrimeNumber.pm:
```diff
@@ -0,0 +1,91 @@
+    }
+    
+    return "None",
+      structured_answer => {
```
I think we're going to need to switch to a "full" template, so we can take advantage of our Info templates ability to hide excess text.

Right now if you entered "primes between 1 and 1,000,000" it would flood the page with text. Using the Info (or possibly list) template. the extra numbers will automatically be hidden until the user clicks a button to expand the text.

moollaza replied:
@abeyang do you have any idea how we should handle these kinds of results, where we list out long lists or short strings?

I feel like the List template isn't ideal because you end up only show 5 or so results, when we could show many more if they items were side-by-side in rows, instead of just 1-per-row

moollaza commented:
@shjnyr thanks for the updates! This is now live on Beta: https://beta.duckduckgo.com/?q=prime+numbers+between+3+and+325&ia=answer

I need to speak with the design team to determine the best way to handle the case of too many results. For example: https://beta.duckduckgo.com/?q=prime+numbers+between+3+and+32500&ia=answer

It may be necessary that we switch this to use full template, likely the Info template.

/cc @abeyang

haojunsui commented:
I have updated my github account info, and also changed it in this script. Can you change the corresponding info in beta server? Sorry for the inconvenience. Thank you.

moollaza commented:
@haojunsui I've updated the Prime Numbers IA Page with your new GitHub username. The Beta server will update when it get's deployed next.

haojunsui commented:
I can't see the description page for this instant answer on beta.DuckDuckGo.com anymore.

moollaza commented:
@haojunsui they get removed once they're merged. This one is slated to go live with the next release though. Should be next week.
